modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/registry/UaParserEnrichment.scala

```diff
@@ -108,7 +106,22 @@ final case class UaParserEnrichment(schemaKey: SchemaKey, parser: Parser) extend
    * @param useragent to extract from. Should be encoded, i.e. not previously decoded.
    * @return the json or the message of the bad row details
    */
-  def extractUserAgent(useragent: String): Either[FailureDetails.EnrichmentFailure, SelfDescribingData[Json]] =
+  def extractUserAgent(useragent: String): F[Either[FailureDetails.EnrichmentFailure, SelfDescribingData[Json]]] =
+    userAgentCache.get(useragent).flatMap {
+      case Some(cachedValue) =>
+        Monad[F].pure(Right(cachedValue))
+      case None =>
+        evaluateUserAgentContext(useragent)
+          .pure[F]
+          .flatTap {
+            case Right(context) =>
+              userAgentCache.put(useragent, context)
+            case Left(_) =>
+              Monad[F].unit
+          }
+    }
+
+  private def evaluateUserAgentContext(useragent: String) =
```

Comment:
Stylistic thing: I have a preference for putting the return type, even when the method is private.

```
private def evaluateUserAgentContext(useragent: String): Either[FailureDetails.EnrichmentFailure, SelfDescribingData[Json]] =
```